### PR TITLE
Improvements for `File`-related configuration inputs tracking

### PR DESCRIPTION
**Pull request by h0tk3y**
Related to https://github.com/gradle/gradle/pull/23266 (fixing https://github.com/gradle/gradle/issues/22656)

### CODE REVIEW

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/inputs/undeclared/GroovyPluginImplementation.groovy [hunk 1]
```diff
@@ -89,14 +91,16 @@ trait GroovyPluginImplementation {
                     }
                     cl("apply")
 
-                    project.tasks.register("thing") { t ->
-                        t.doLast {
-                            value = ${read.groovyExpression}
-                            println("task = " + value)
+                    if (!project.tasks.names.contains("thing")) {
```
Comment (mlopatkin) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/inputs/undeclared/GroovyPluginImplementation.groovy:94]: I'm not sure if introducing discrepancies into plugin implementations is a good idea. Maybe we can get away with an ad hoc implementation of the template in the test itself? Then you can drop all unrelated checks for INSTANCE, etc.

File: subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/inputs/undeclared/UndeclaredFileInputsIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import org.gradle.configurationcache.AbstractConfigurationCacheIntegrationTest
+import org.gradle.test.fixtures.file.TestFile
+
+import static org.gradle.configurationcache.inputs.undeclared.FileUtils.testFilePath
+import static org.hamcrest.CoreMatchers.containsString
+import static org.hamcrest.CoreMatchers.startsWith
+import static org.hamcrest.Matchers.allOf
+
+class UndeclaredFileInputsIntegrationTest extends AbstractConfigurationCacheIntegrationTest implements GroovyPluginImplementation {
+    @NotYetImplemented
```
Comment (mlopatkin) [inline subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/inputs/undeclared/UndeclaredFileInputsIntegrationTest.groovy:29]: I think we prefer to use `@ToBeImplemented` instead to test for a specific kind of failure.

**Review by mlopatkin — CHANGES_REQUESTED** [review]
Looks pretty good! I have a few concerns about tests, though.

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Comment by h0tk3y** [comment]
@bot-gradle test and merge

**Pull request merged**